### DISCUSSION TITLE
(CLI) Oh boy oh boy oh boy!

### DISCUSSION THREAD

**Discussion by BMCwebdev** [facebook/stylex#583] [General]
Saw a certain branch get merged this morning! I will be giving the CLI a spin as soon as you publish it. Hoping to smooth over some of the more bumpy parts of using StyleX in Next app, especially when running dev. Yea!
🎉

**Comment by zaydek** [comment]
What does the CLI do?

**Reply by Jta26** [reply]
The CLI is an alternative to integrating stylex with an existing build system. It's an escape hatch for non-babel stacks.  In cases where your existing build stack doesn't use babel, we made a standalone wrapper around the babel plugin that can act as a "pre-build" step to compile all your stylex styles.

**Reply by zaydek** [reply]
Oh interesting, so like if I use esbuild (assuming I'm not using https://www.npmjs.com/package/@stylexjs/esbuild-plugin), you're saying in theory I can still use StyleX in that dev server?

**Reply by Jta26** [reply]
Yes, this is the idea.

**Reply by javascripter** [reply]
@Jta26 
Do you think it makes sense to have an option to output `stylex.bundle.css` without emitting any js files, similar to Tailwind CLI?

My idea is:
1. Frameworks like Next.js and Expo allow configuring babel plugins in babel.config.js or babel-loader via webpack/Turbopack, so use`'@stylexjs/babel-plugin'` to transform the source code only (with runtimeInjection: false) and discard metadata here
2. The CLI sidesteps the bundler completely and transforms the source code separately (basically redundant work), but actually collects the metadata and writing the CSS to a specified path

Something like this:
```bash
stylex -i ./src -o src/stylex.bundle.css --watch
```

If this is feasible, then
1. You don't need to gitignore `src` directory, so tsconfig.json, aliases, module resolutions etc work without any additional change except for the standard babel.config.js setup
2. Since no JS files will be written to src, you don't have to worry about timing issues (if Next.js dev server starts up before src/ directory exists, it exists with an error).

The downsides are:
a. redundant transform work in the CLI
b. potential mismatch of output if babel configs aren't in sync between the app and CLI

Or, are there other downsides to this approach, like treeshaking?

**Reply by nmn** [reply]
@javascripter that may work in some scenarios.

The initial version of the CLI was made primarily for Next.js and SWC so it chooses to compile the entire folder. I could see an option being added to make it work the way you described.
